rfcs/0000-kuadrant-ci-cd.md

```diff
@@ -0,0 +1,835 @@
+needs them.
+
+# Motivation
+[motivation]: #motivation
+
+The current process is not only slow, but also involves manual steps that are prone to human error. Its implementation
+is also not very flexible, and it's not easy to add new steps or change the existing ones without replicating the same
+code in different repositories. It also involves a convolution of different tools and services that are not easy to
+maintain and that are not very well integrated with each other.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Current process
+
+<details>
+<summary>Summary of the current process</summary>
+
+Before we dive into the new process, let's take a look at the current one. The current process is based on the
+independent build and release of each service (Authorino and Limitador), their respective Operators and the WasmShim
+for then building and releasing the Kuadrant Operator. It's important that we follow this order because the
+Kuadrant Operator depends on the WasmShim, and the Authorino and Limitador Operators. The process is as follows:
+
+### Authorino
+A particularity of Authorino is that it follows the [Controller Pattern](https://kubernetes.io/docs/concepts/architecture/controller/#controller-pattern),
+it's implemented as a Kubernetes controller thanks to the [Operator SDK](https://sdk.operatorframework.io/) and
+[Kustomize](https://kustomize.io/).
+
+#### Artifacts
+The deliverable artifacts are the [Authorino image](https://quay.io/repository/kuadrant/authorino)
+and its [manifests](https://github.com/Kuadrant/authorino/blob/main/install/kustomization.yaml), the AuthConfig CRD and
+role definitions.
+
+#### Build / Release
+The build process is triggered by a particular GitHub Action that runs on demand (workflow_dispatch) named "Build and push image",
+and to build an image version “v0.X.Y” of Authorino, the following steps are needed:
+
+1. Pick a <git-ref> (SHA-1) as source.
+Create a new tag and named release “v0.X.Y”. Make sure to write change notes highlighting all the new features, bug fixes,
+enhancements, etc ([example](https://github.com/Kuadrant/authorino/releases/tag/v0.9.0)).
+
+2. Run the GHA ‘Build and push images’ for the “v0.X.Y” tag. This will cause a new image to be built and pushed to quay.io/kuadrant/authorino.
+
+3. Create the release and release notes on [Github](https://github.com/Kuadrant/authorino/releases/new) using the tag from above, named: `Authorino vM.m.d`
+
+Notes:
+PRs merged to the main branch of Authorino cause a new image to be built (GH Action) and pushed automatically to
+quay.io/kuadrant/authorino:<git-ref> – the quay.io/kuadrant/authorino:latest tag is also moved to match the latest <git-ref>.
+
+### Authorino Operator
+The Authorino operator is similar to its operand, it fully follows the [Operator Pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/)
```

Comment:
Authorino Operator uses nowadays a mix of all options below (1, 2 and 3).

There's is a `build.yaml` file (option 1) that is created by a script/make target (option 2), and committed to each release branch.

The file contains the version of the operand the operator is bound to (in the form of a container image ref) and the value of the `replaces` directive to be used in the operator bundle manifests. I.e. all info it needs to prepare a new release.

E.g. https://github.com/Kuadrant/authorino-operator/blob/release-v0.10.0/build.yaml

The process of creating the release branch with its build file, generating the operator bundle, and publishing the GitHub release notes is triggered in GitHub, by executing the _Release operator_ action (option 3).